nym-vpn-core/nym-vpn-lib/src/platform/mod.rs

```diff
@@ -91,21 +100,36 @@ async fn _async_run_vpn(vpn: SpecificVpn) -> Result<(Arc<Notify>, NymVpnHandle),
     let mut handle = spawn_nym_vpn(vpn)?;
     debug!("spawned vpn handle");
 
-    match handle
+    let status_update = handle
         .vpn_status_rx
         .next()
         .await
-        .ok_or(crate::Error::NotStarted)?
-        .downcast_ref::<TaskStatus>()
-        .ok_or(crate::Error::NotStarted)?
-    {
-        TaskStatus::Ready => debug!("Started Nym VPN"),
-        TaskStatus::ReadyWithGateway(gateway) => {
-            debug!("Started Nym VPN: connected to {gateway}");
-            set_listener_status(TunStatus::Up);
+        .ok_or(crate::Error::NotStarted)?;
+    if let Some(message) = status_update.downcast_ref::<TaskStatus>() {
+        match message {
+            TaskStatus::Ready => debug!("Started Nym VPN"),
+            TaskStatus::ReadyWithGateway(gateway) => {
+                debug!("Started Nym VPN: connected to {gateway}");
+                set_listener_status(StatusEvent::Tun(TunStatus::Up));
+            }
         }
     }
 
+    if let Some(message) = status_update.downcast_ref::<BandwidthStatusMessage>() {
+        set_listener_status(StatusEvent::Bandwidth(message.into()))
+    }
+
+    if let Some(message) = status_update
+        .downcast_ref::<ConnectionMonitorStatus>()
+        .cloned()
+    {
+        set_listener_status(StatusEvent::Connection(message.into()))
+    }
+
+    if let Some(message) = status_update.downcast_ref::<NymVpnStatusMessage>().cloned() {
+        set_listener_status(StatusEvent::NymVpn(message.into()))
+    }
+
```

Comment:
It looks like this block above only checks for the first event? One level up in the call stack it's spawning a listener for the exit message, but I'm thinking there should also be one for status messages there?

Reply:
I think I've fixed this. I am now getting consistent status updates, however, sometimes when I try to connect the app crashes:



Issue with this PR: 

453 libnymvpn               net.nymtech.nymvpn                   D  nym_vpn_lib::routing: Creating tun device
2024-07-30 01:27:32.340 18038-18453 libnymvpn               net.nymtech.nymvpn                   D  jni::wrapper::java_vm::vm: Attached daemon thread tokio-runtime-worker (ThreadId(15)). 8 threads attached
2024-07-30 01:27:32.341 18038-18453 libnymvpn               net.nymtech.nymvpn                   D  jni::wrapper::java_vm::vm: Detached thread tokio-runtime-worker (ThreadId(15)). 7 threads remain attached
2024-07-30 01:27:32.344 18038-18453 libnymvpn               net.nymtech.nymvpn                   D  nym_task::manager: [TaskClient-nym_vpn_lib-mixnet_client_main] the task client is getting dropped: this is expected during client shutdown
2024-07-30 01:27:32.345 18038-18174 libnymvpn               net.nymtech.nymvpn                   D  rustls::common_state: Sending warning alert CloseNotify
--------- beginning of crash
2024-07-30 01:27:32.353 18038-18174 libnymvpn               net.nymtech.nymvpn                   D  rustls::common_state: Sending warning alert CloseNotify
2024-07-30 01:27:32.354 18038-18453 libc                    net.nymtech.nymvpn                   A  Fatal signal 6 (SIGABRT), code -1 (SI_QUEUE) in tid 18453 (Thread-105), pid 18038 (.nymtech.nymvpn)
2024-07-30 01:27:32.426 18038-18493 libnymvpn               net.nymtech.nymvpn                   I  nym_client_core::client::packet_statistics_control: down: 3.21 kiB/s, up: 4.71 kiB/s (cover down: 821.33 B/s, cover up: 5.89 kiB/s)
2024-07-30 01:27:32.427 18038-18493 libnymvpn               net.nymtech.nymvpn                   D  nym_client_core::client::packet_statistics_control: RX: 2.5 mixpkt/s, 4.01 kiB/s (real: 3.21 kiB/s, acks: 41.99 B/s), TX: 4.5 mixpkt/s, 10.60 kiB/s (real: 4.71 kiB/s)
2024-07-30 01:27:32.427 18038-18493 libnymvpn               net.nymtech.nymvpn                   D  nym_client_core::client::packet_statistics_control: packets sent: 9 (real: 4, cover: 5, retransmissions: 0)
2024-07-30 01:27:32.427 18038-18493 libnymvpn               net.nymtech.nymvpn                   D  nym_client_core::client::packet_statistics_control: packets received: 5, (real: 4, cover: 1, acks: 4, acks for cover: 0)
2024-07-30 01:27:32.825 18862-18862 DEBUG                   crash_dump64                         A  Cmdline: net.nymtech.nymvpn
2024-07-30 01:27:32.825 18862-18862 DEBUG                   crash_dump64                         A  pid: 18038, tid: 18453, name: Thread-105  >>> net.nymtech.nymvpn <<<
2024-07-30 01:27:32.825 18862-18862 DEBUG                   crash_dump64                         A        #01 pc 000000000156dd58  /data/app/~~Crwu6XktK_7OwNKgmNbknw==/net.nymtech.nymvpn-HSmfK7YF9N_d0yIj8SeFtQ==/base.apk (offset 0x32d0000)
2024-07-30 01:27:32.825 18862-18862 DEBUG                   crash_dump64                         A        #02 pc 00000000015692d4  /data/app/~~Crwu6XktK_7OwNKgmNbknw==/net.nymtech.nymvpn-HSmfK7YF9N_d0yIj8SeFtQ==/base.apk (offset 0x32d0000)
2024-07-30 01:27:32.825 18862-18862 DEBUG                   crash_dump64                         A        #03 pc 0000000001568f20  /data/app/~~Crwu6XktK_7OwNKgmNbknw==/net.nymtech.nymvpn-HSmfK7YF9N_d0yIj8SeFtQ==/base.apk (offset 0x32d0000)
2024-07-30 01:27:32.825 18862-18862 DEBUG                   crash_dump64                         A        #04 pc 0000000001567be8  /data/app/~~Crwu6XktK_7OwNKgmNbknw==/net.nymtech.nymvpn-HSmfK7YF9N_d0yIj8SeFtQ==/base.apk (offset 0x32d0000)
2024-07-30 01:27:32.825 18862-18862 DEBUG                   crash_dump64                         A        #05 pc 0000000001568cc0  /data/app/~~Crwu6XktK_7OwNKgmNbknw==/net.nymtech.nymvpn-HSmfK7YF9N_d0yIj8SeFtQ==/base.apk (offset 0x32d0000)
2024-07-30 01:27:32.825 18862-18862 DEBUG                   crash_dump64                         A        #06 pc 000000000157f2d0  /data/app/~~Crwu6XktK_7OwNKgmNbknw==/net.nymtech.nymvpn-HSmfK7YF9N_d0yIj8SeFtQ==/base.apk (offset 0x32d0000)
2024-07-30 01:27:32.825 18862-18862 DEBUG                   crash_dump64                         A        #07 pc 000000000157f384  /data/app/~~Crwu6XktK_7OwNKgmNbknw==/net.nymtech.nymvpn-HSmfK7YF9N_d0yIj8SeFtQ==/base.apk (offset 0x32d0000)
2024-07-30 01:27:32.825 18862-18862 DEBUG                   crash_dump64                         A        #08 pc 000000000157f44c  /data/app/~~Crwu6XktK_7OwNKgmNbknw==/net.nymtech.nymvpn-HSmfK7YF9N_d0yIj8SeFtQ==/base.apk (offset 0x32d0000)
2024-07-30 01:27:32.825 18862-18862 DEBUG                   crash_dump64                         A        #09 pc 00000000007f5454  /data/app/~~Crwu6XktK_7OwNKgmNbknw==/net.nymtech.nymvpn-HSmfK7YF9N_d0yIj8SeFtQ==/base.apk (offset 0x32d0000)
2024-07-30 01:27:32.825 18862-18862 DEBUG                   crash_dump64                         A        #10 pc 00000000007ce294  /data/app/~~Crwu6XktK_7OwNKgmNbknw==/net.nymtech.nymvpn-HSmfK7YF9N_d0yIj8SeFtQ==/base.apk (offset 0x32d0000)
2024-07-30 01:27:32.825 18862-18862 DEBUG                   crash_dump64                         A        #11 pc 0000000000a73664  /data/app/~~Crwu6XktK_7OwNKgmNbknw==/net.nymtech.nymvpn-HSmfK7YF9N_d0yIj8SeFtQ==/base.apk (offset 0x32d0000)
2024-07-30 01:27:32.825 18862-18862 DEBUG                   crash_dump64                         A        #12 pc 0000000001548570  /data/app/~~Crwu6XktK_7OwNKgmNbknw==/net.nymtech.nymvpn-HSmfK7YF9N_d0yIj8SeFtQ==/base.apk (offset 0x32d0000)
2024-07-30 01:27:32.825 18862-18862 DEBUG                   crash_dump64                         A        #13 pc 0000000001547560  /data/app/~~Crwu6XktK_7OwNKgmNbknw==/net.nymtech.nymvpn-HSmfK7YF9N_d0yIj8SeFtQ==/base.apk (offset 0x32d0000)
2024-07-30 01:27:32.825 18862-18862 DEBUG                   crash_dump64                         A        #14 pc 0000000001543c90  /data/app/~~Crwu6XktK_7OwNKgmNbknw==/net.nymtech.nymvpn-HSmfK7YF9N_d0yIj8SeFtQ==/base.apk (offset 0x32d0000)
2024-07-30 01:27:32.825 18862-18862 DEBUG                   crash_dump64                         A        #15 pc 00000000015483e0  /data/app/~~Crwu6XktK_7OwNKgmNbknw==/net.nymtech.nymvpn-HSmfK7YF9N_d0yIj8SeFtQ==/base.apk (offset 0x32d0000)
2024-07-30 01:27:32.825 18862-18862 DEBUG                   crash_dump64                         A        #16 pc 00000000007d0cbc  /data/app/~~Crwu6XktK_7OwNKgmNbknw==/net.nymtech.nymvpn-HSmfK7YF9N_d0yIj8SeFtQ==/base.apk (offset 0x32d0000)
2024-07-30 01:27:32.825 18862-18862 DEBUG                   crash_dump64                         A        #17 pc 0000000000a75f88  /data/app/~~Crwu6XktK_7OwNKgmNbknw==/net.nymtech.nymvpn-HSmfK7YF9N_d0yIj8SeFtQ==/base.apk (offset 0x32d0000)
2024-07-30 01:27:32.825 18862-18862 DEBUG                   crash_dump64                         A        #18 pc 00000000015355e8  /data/app/~~Crwu6XktK_7OwNKgmNbknw==/net.nymtech.nymvpn-HSmfK7YF9N_d0yIj8SeFtQ==/base.apk (offset 0x32d0000)
2024-07-30 01:27:32.825 18862-18862 DEBUG                   crash_dump64                         A        #19 pc 000000000154ad24  /data/app/~~Crwu6XktK_7OwNKgmNbknw==/net.nymtech.nymvpn-HSmfK7YF9N_d0yIj8SeFtQ==/base.apk (offset 0x32d0000)
2024-07-30 01:27:32.825 18862-18862 DEBUG                   crash_dump64                         A        #20 pc 000000000152b6a8  /data/app/~~Crwu6XktK_7OwNKgmNbknw==/net.nymtech.nymvpn-HSmfK7YF9N_d0yIj8SeFtQ==/base.apk (offset 0x32d0000)
2024-07-30 01:27:32.825 18862-18862 DEBUG                   crash_dump64                         A        #21 pc 000000000156cdf0  /data/app/~~Crwu6XktK_7OwNKgmNbknw==/net.nymtech.nymvpn-HSmfK7YF9N_d0yIj8SeFtQ==/base.apk (offset 0x32d0000)